### PR TITLE
Migrate ziplist.c unit tests to new framework

### PR DESCRIPTION
**Pull request by hallmason17**
Issue #428.

Moved the SERVER_TEST block from ziplist.c into unit tests in test_ziplist.c. I left the benchmark related tasks alone in their own test, as I am not sure what to do with them. 

Some of my assertions are a little vague/useless, but I will try to refine them.

image

### CODE REVIEW

**Comment by hallmason17** [comment]
Looks like I may also need to add a test that does pop() and another for createIntList(), since I did not add that back.

File: src/unit/test_ziplist.c [hunk 1]
```diff
@@ -0,0 +1,886 @@
+#include <sys/time.h>
+#include "../ziplist.c"
+#include "../adlist.h"
+#include "test_help.h"
+
+#define debug(f, ...) { if (DEBUG) printf(f, __VA_ARGS__); }
```
Comment (madolson) [inline src/unit/test_ziplist.c]:
```suggestion
```

Seems like this was always unused.

File: src/unit/test_ziplist.c [hunk 2]
```diff
@@ -0,0 +1,886 @@
...
+
+    zfree(zl);
+    return 0;
+}
+
+int test_ziplistBenchmarks(int argc, char **argv, int flags) {
+    if (argc >= 4)
+        srand(atoi(argv[3]));
+    int accurate = (flags & UNIT_TEST_ACCURATE);
+    printf("Stress with variable ziplist size:\n");
```
Comment (madolson) [inline src/unit/test_ziplist.c]:
```suggestion
    TEST_PRINT_INFO("Stress with variable ziplist size:");
```
Ideally let's omit any direct printf call. We can better capture it in the future this way.

File: src/unit/test_ziplist.c [hunk 3]
```diff
@@ -0,0 +1,886 @@
...
+            zl = ziplistPush(zl, (unsigned char *) data, ZIP_BIG_PREVLEN - 4, ZIPLIST_TAIL);
+        }
+        unsigned long long start = usec();
+        zl = ziplistPush(zl, (unsigned char *) data, ZIP_BIG_PREVLEN - 3, ZIPLIST_HEAD);
+        printf("Done. usec=%lld\n\n", usec() - start);
+        zfree(zl);
+    }
+
+
+    printf("Benchmarks done\n");
```
Comment (madolson) [inline src/unit/test_ziplist.c]:
```suggestion
```
No longer really needed, as the test itself will print it's done.

File: src/unit/test_ziplist.c [hunk 4]
```diff
@@ -0,0 +1,886 @@
...
+    unsigned int vlen;
+    long long vlong = 0;
+
+    p = ziplistIndex(zl, where == ZIPLIST_HEAD ? 0 : -1);
+    if (ziplistGet(p, &vstr, &vlen, &vlong)) {
+        if (where == ZIPLIST_HEAD)
+            printf("Pop head: ");
+        else
+            printf("Pop tail: ");
+
```
Comment (madolson) [inline src/unit/test_ziplist.c]:
```suggestion
```

These, for the most part, just spam our output in tests. So I would omit them:

File: src/unit/test_ziplist.c [hunk 5]
```diff
@@ -0,0 +1,886 @@
...
+        if (vstr) {
+            if (vlen && fwrite(vstr, vlen, 1, stdout) == 0) perror("fwrite");
+        } else {
+            printf("%lld", vlong);
+        }
+
+        printf("\n");
```
Comment (madolson) [inline src/unit/test_ziplist.c]:
```suggestion
        }
```

File: src/unit/test_ziplist.c [hunk 6]
```diff
@@ -0,0 +1,886 @@
...
+    p = ziplistIndex(zl, 0);
+    while (ziplistGet(p, &entry, &elen, &value)) {
+        TEST_ASSERT(p != NULL);
+        p = ziplistNext(zl, p);
+    }
+    zfree(zl);
+    return 0;
+}
+
+int test_ziplistInterateThroughListFrom1ToEnd(int argc, char **argv, int flags) {
```
Comment (madolson) [inline src/unit/test_ziplist.c]:
```suggestion
int test_ziplistIterateThroughListFrom1ToEnd(int argc, char **argv, int flags) {
```

File: src/unit/test_ziplist.c [hunk 7]
```diff
@@ -0,0 +1,886 @@
...
+    p = ziplistIndex(zl, 1);
+    while (ziplistGet(p, &entry, &elen, &value)) {
+        TEST_ASSERT(p != NULL);
+        p = ziplistNext(zl, p);
+    }
+    zfree(zl);
+    return 0;
+}
+
+int test_ziplistInterateThroughListFrom2ToEnd(int argc, char **argv, int flags) {
```
Comment (madolson) [inline src/unit/test_ziplist.c]:
```suggestion
int test_ziplistIterateThroughListFrom2ToEnd(int argc, char **argv, int flags) {
```

File: src/unit/test_ziplist.c [hunk 8]
```diff
@@ -0,0 +1,886 @@
...
+    p = ziplistIndex(zl, 2);
+    while (ziplistGet(p, &entry, &elen, &value)) {
+        TEST_ASSERT(p != NULL);
+        p = ziplistNext(zl, p);
+    }
+    zfree(zl);
+    return 0;
+}
+
+int test_ziplistInterateThroughStartOutOfRange(int argc, char **argv, int flags) {
```
Comment (madolson) [inline src/unit/test_ziplist.c]:
```suggestion
int test_ziplistIterateThroughStartOutOfRange(int argc, char **argv, int flags) {
```

File: src/ziplist.c [hunk 9]
```diff
@@ -1687,982 +1687,3 @@ unsigned int ziplistRandomPairsUnique(unsigned char *zl, unsigned int count, zip
...
-    ziplistRepr(zl);
-
-    zl = pop(zl,ZIPLIST_HEAD);
-    ziplistRepr(zl);
-
-    zl = pop(zl,ZIPLIST_TAIL);
-    ziplistRepr(zl);
-
-    zl = pop(zl,ZIPLIST_TAIL);
-    ziplistRepr(zl);
```
Comment (madolson) [inline src/ziplist.c:1881]: It looks like this code got dropped and wasn't added to a new new test.

**Review by madolson — COMMENTED** [review]
Looks pretty good, thanks for contributing!

**Comment by hallmason17** [comment]
Yikes, I'm not sure how to fix this signing issue. I tried the rebasing instructions, but the result did not look right...

**Comment by madolson** [comment]
The best way is probably to do a `git rebase -i HEAD~14`, squash all the commit together, sign just the last one, and force push over what you have.

**Review by madolson — APPROVED** [review]
LGTM. Ran some more tests, but it looks like it was hitting the issue that we fixed with zmalloc, so seems okay to follow up on that in mainline if it causes an issue.

**Comment by madolson** [comment]
All the updates LGTM feel free to update them further if you like.

File: src/unit/test_ziplist.c [hunk 10]
```diff
@@ -0,0 +1,1049 @@
...
+        srand(atoi(argv[3]));
+    zl = createList(); /* "hello", "foo", "quux", "1024" */
+
+    p = ziplistIndex(zl, 0);
+    TEST_ASSERT(ziplistCompare(p, (unsigned char *)"hello", 5));
+    int orig_len = ziplistLen(zl);
+
+    zl = ziplistDeleteRange(zl, 0, 1);
+
+    TEST_ASSERT(ziplistCompare(p, (unsigned char *)"foo", 3));
```
Comment (madolson) [inline src/unit/test_ziplist.c:364]: I believe p is getting invalidated here because of the ziplistDeleteRange. I think you need to reinitialize it.

Comment (hallmason17) [reply src/unit/test_ziplist.c:364]: Looks like Ping may have beat me to it with #537. Sorry about that.

Comment (madolson) [reply src/unit/test_ziplist.c:364]: No worries I tried to run the daily tests before merging it, and there was another known issue so I merged it anyways.  :/

**Pull request merged**